### DISCUSSION TITLE
State in not singlet in SA-CASSCF

### DISCUSSION THREAD

**Discussion by JunboLu** [pyscf/pyscf#2629] [Q&A]
Dear developers:
  I am using SA-CASSCF method to calculate Eu3+ cation. My main input is:
solver = mcscf.CASSCF(mf, 7, 6)
solver.chkfile = 'ln_cas.chk'

solvers = []
solver1 = fci.direct_spin0.FCI(mol1)
solver1.spin = 0
solver1 = fci.addons.fix_spin(solver1,ss=(float(solver1.spin)/2.0)*(float(solver1.spin)/2.0+1.0),shift=2.0)
solver1.pspace_size = 10000
#solver1.max_space = 400
#solver1.max_cycle = 30
solver1.nroots = 100
solvers.append(solver1)

solver2 = fci.direct_spin1.FCI(mol1)
solver2.spin = 2
solver2 = fci.addons.fix_spin(solver2,ss=(float(solver2.spin)/2.0)*(float(solver2.spin)/2.0+1.0),shift=2.0)
solver2.nroots = 100
solvers.append(solver2)

solver3 = fci.direct_spin1.FCI(mol1)
solver3.spin = 4
solver3 = fci.addons.fix_spin(solver3,ss=(float(solver3.spin)/2.0)*(float(solver3.spin)/2.0+1.0),shift=2.0)
solver3.nroots = 100
solvers.append(solver3)

solver4 = fci.direct_spin1.FCI(mol1)
solver4.spin = 6
solver4 = fci.addons.fix_spin(solver4,ss=(float(solver4.spin)/2.0)*(float(solver4.spin)/2.0+1.0),shift=2.0)
solver4.nroots = 7
solvers.append(solver4)

#mcscf.state_average_mix_(solver, solvers, np.ones(solver2.nroots+solver3.nroots)/(solver2.nroots+solver3.nroots))
mcscf.state_average_mix_(solver, solvers, np.ones(solver1.nroots+solver2.nroots+solver3.nroots+solver4.nroots)/(solver1.nroots+solver2.nroots+solver3.nroots+solver4.nroots))

caslst = [55, 56, 57, 58, 59, 60, 61]
mo1 = mcscf.addons.sort_mo(solver,mo_coeff,caslst,1)

solver.run(mo1)

But I get an error:
Traceback (most recent call last):
  File "/mnt/sg001/home/fz_sustech_ljb/WORK/lujb/tetra_effect/LnCl3/CASSCF_NEVPT2_SOC/EuCl3/SA-CASSCF/input.py", line 167, in <module>
    solver.run(mo1)
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/lib/misc.py", line 512, in run
    self.kernel(*args)
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/mcscf/mc1step.py", line 812, in kernel
    _kern(self, mo_coeff,
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/mcscf/mc1step.py", line 349, in kernel
    e_tot, e_cas, fcivec = casscf.casci(mo, ci0, eris, log, locals())
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/mcscf/mc1step.py", line 834, in casci
    e_tot, e_cas, fcivec = casci.kernel(fcasci, mo_coeff, ci0, log)
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/mcscf/casci.py", line 546, in kernel
    e_tot, fcivec = casci.fcisolver.kernel(h1eff, eri_cas, ncas, nelecas,
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/mcscf/addons.py", line 1265, in kernel
    e, c = solver.kernel(h1, h2, norb, self._get_nelec(solver, nelec), c0,
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/fci/direct_spin0.py", line 367, in kernel
    kernel_ms0(self, h1e, eri, norb, nelec, ci0, None,
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/fci/direct_spin0.py", line 331, in kernel_ms0
    return e+ecore, [_check_(ci.reshape(na,na)) for ci in c]
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/fci/direct_spin0.py", line 331, in <listcomp>
    return e+ecore, [_check_(ci.reshape(na,na)) for ci in c]
  File "/mnt/sg001/home/fz_sustech_ljb/bin/pyscf-2.0.1/pyscf/fci/direct_spin0.py", line 340, in _check_
    raise ValueError('State not singlet %g' % (norm - 1))
ValueError: State not singlet -2.02771e-05

I try to increase max_space and pspace_size ( > 10000 ), the error still exists. Does anyone have other suggestions? I would be very appreciated!
  With my best regards,
  Junbo

**Comment by MatthewRHermes** [comment]
As the commentary in the source code itself states:

```
direct_spin0 solver is specified for singlet state. However, calling this
solver sometimes ends up with the error "State not singlet x.xxxxxxe-06" due
to numerical issues. Calling direct_spin1 for singlet state is slightly
slower but more robust than direct_spin0 especially when combining to energy
penalty method (:func:`fix_spin_`)
```

What happens when you just use direct_spin1 for everything? Are you able to get the right spin states then?

I'm not sure about the robustness of the spin-penalty method when you are trying to get many dozens of roots. Logically, I would think this would require the shift be so large that very first root of incorrect spin is rendered higher energy than the 100th root of the correct spin, which would often be a very large shift, which would result in very slow, difficult convergence.

You could use [my](https://github.com/MatthewRHermes/mrh) csf_solver, which doesn't rely on a penalty:

```
from mrh.my_pyscf.fci import csf_solver

solver1 = csf_solver (mol1, smult=1).set (spin=0, nroots=100)
solver2 = csf_solver (mol1, smult=3).set (spin=2, nroots=100)
# etc
```

but beware that it is memory intensive and not up-to-date with PySCF developments in terms of very high-symmetry systems (such as single atoms).

**Comment by JunboLu** [comment]
Thank you in advance!
I try to use spin1, there is something wrong with the spin state:

![72b200dc0fa96846db0ddb00f356fa1](https://github.com/user-attachments/assets/1cdb3579-500f-4d2a-8c75-3331953c4fa3)

I calculate 100 singlet states. But from No. 69 state, there is triplet, not the singlet.

**Comment by MatthewRHermes** [comment]
What is the lowest non-singlet energy of this molecule (i.e., from the other solvers)? As I said above, I think if you want 100 singlets, then you need to set the shift parameter to no less than the energy difference between the 100th singlet and the lowest-energy non-singlet.
